Typed events

arinwt opened the pull request:
Draft PR for getting stronger typing on TypedEventEmitters.

See `common/lib/common-utils/src/test/test-d/typedEventEmitter.test-d.ts` for examples on how it should be used.

It gets strong typing for both emitting and listening to events. It is just a direct typecast over EventEmitter, so it should have no runtime cost.

It works with extended generic events as well, but the drawback is that we lose a little bit of type specificity with the "newListener" and "removeListener" event typings.

The core of the change is that TypedEventEmitter is a hard cast of the EventEmitter constructor to IEventEmitter implementing the IEventProvider interface:
```ts
export const TypedEventEmitter = EventEmitter as unknown as new <TEvents>() => IEventEmitter<TEvents>;
```
`IEventProvider` only provides listening functions, and `IEventEmitter` is closer to the full typing for an EventEmitter, including `emit` and other functions, as well as the "newListener", "removeListener", and "error" events.

@anthony-murphy - If it looks good, I can start propagating the change. I would love to know a good complex place to start to ensure it's going to be sufficient.

anthony-murphy commented:
not sure what testing you've done here. but i'll frequently put something like this somewhere easily accessible, like container-utils, and move everything over to it, and make sure things compile, and that intelligence is behaving as expected. Also, this is a breaking change, so would need to come in the next minor(major) release, to avoid that i'd suggest adding this in parallel to the exiting implementation, rather than replacing it.

anthony-murphy commented:
I actually think the only thing using typed events above the client is quorum, which i'd like to decouple from protocol-defs, not sure when, so maybe this gets checked into the client layer somewhere, we stop using the current one expect for quorum, and deal with that when we deal with quorum

arinwt commented:
The testing is just tsd type testing, to make sure things transpile as expected and TypeScript gives me the types I expect.

I'm on board with that, I'll see what it looks like in one of the runtime or container utils first.

Commit:
Add typed event emitter to runtime utils

Commit:
Fix sequence deps

Commit:
Merge branch 'main' into typed-events-2

Commit:
Sync tsd dep version

Commit:
Sync versions of tsd

Commit:
Add EventThis placeholder type for this

Commit:
Add separation between functions that replace EventThis and arrays which do not

Commit:
Update all dds packages

Commit:
Update intelligence runner agent to use typed events

Commit:
Add brief comment

arinwt commented:
@anthony-murphy - I went ahead and added the type contracts to runtime-definitions, and TypedEventEmitter to runtime-utils.

I updated all packages in `packages/runtime` and `packages/dds` to use this implementation.

I also added "this" replacement similar to how you did. In order to make the types work more cleanly with complex types (like Jsonable), I made the following distinction:
```ts
interface ITestEvents {
    literal: [a: number, b: EventThis];
    replaced(a: number, b: EventThis): EventThis; // returning EventThis isn't necessary right now
}
function test(emitter: IEventProvider<ITestEvents>) {
    emitter.on("literal", (a, b) => {
        assert(typeof a === "string");
        b.on("literal", ...); // this fails, because b is "EventThis" (unique symbol)
    });
    emitter.on("replaced", (a, b) => {
        assert(typeof a === "string");
        b.on("literal", ...); // this works, because b is replaced to be "this" of IEventProvider implementer
    });
}
```

So basically you can use either an array `[]` (preferably named tuple) to designate the literal types, or you can use a function, which will perform type replacement. The benefit of keeping the array type around is that some complex types won't work cleanly when checking against `EventThis` (`Serializable<T>` for example).

The only drawback I've found so far (which I think I get) is that when emitting a `this` type in a generic context, it requires being explicit about the types. This can be seen in the types tests:
```ts
this.emit<"testThis", IBaseEvents["testThis"]>("testThis", this); // works
this.emit("testThis", this); // fails, it's not sure if it can cast this. I think because extended generic type could be different, or at least it thinks so.
```

Meanwhile for simpler cases, it works fine when in outermost contexts:
```ts
this.emit("this", 2, this); // works
```

anthony-murphy commented on packages/runtime/runtime-definitions/src/events.ts:
```diff
@@ -0,0 +1,54 @@
+ * Licensed under the MIT License.
+ */
+
+export const eventThis = Symbol("EventThis");
```
>Symbol [](http://example.com/codeflow?start=25&length=6)

why use a symbol? this will actually create an artifact that exists at runtime.

arinwt replied:
I did that to prevent collisions, I figured the cost was small, but I'm not really opposed to just picking an extremely unlikely collision as well.

arinwt replied:
Oh I think I can achieve the same thing with `declare const eventThis: unique symbol;`.

Commit:
Use unique symbol type as placeholder instead of actual runtime const

Commit:
Merge branch 'main' into typed-events

Commit:
Add missing ink events

arinwt commented on packages/dds/ink/src/interfaces.ts:
```diff
@@ -56,8 +56,9 @@ export interface IColor {
-    (event: "stylus", listener: (operation: IStylusOperation) => void);
-    (event: "clear", listener: () => void);
+    stylus: [operation: IStylusOperation];
+    clear: [operation: IClearOperation];
+    createStroke: [operation: ICreateStrokeOperation];
```
These event typings were mismatched/wrong before, and gave an error in the `emit` calls with my changes. Corrected to include the operation in `clear` and added the `createStroke` event.

arinwt commented on packages/dds/map/src/directory.ts:
```diff
@@ -1189,7 +1192,7 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
         }
         this.clearExceptPendingKeys();
-        this.directory.emit("clear", local, op, this.directory);
+        this.directory.emit("clear", local, message, this.directory);
```
Note that this was actually mismatched/wrong before, and was detected by adding typings to `emit`.
Before it was sending `IDirectoryClearOperation` instead of the actual `ISequencedDocumentMessage` as indicated by the typing.

arinwt commented on packages/dds/map/src/interfaces.ts:
```diff
@@ -144,16 +134,17 @@ export interface IDirectoryValueChanged extends IValueChanged {
-    (event: "valueChanged", listener: (
-        changed: IDirectoryValueChanged,
+    valueChanged(
+        changed: IValueChanged,
```
Note that this was wrong/mismatched before. In the emit calls, it doesn't send an `IDirectoryValueChanged`, but rather an `IValueChanged`. We could change the emit calls or the typing. Not sure, but they should be synced.

arinwt replied:
Because I did it this way, I needed to change a bunch of downstream code to dependon `IValueChanged` instead of `IDirectoryValueChanged`. Curious what is correct here? - for example the react package.

ChumpChief replied:
Oh sorry, I just fixed this today: https://github.com/microsoft/FluidFramework/pull/6397

arinwt commented on packages/dds/map/src/mapKernel.ts:
```diff
@@ -481,7 +481,7 @@ export class MapKernel {
         this.data.set(key, value);
         const event: IValueChanged = { key, previousValue };
-        this.eventEmitter.emit("valueChanged", event, local, op, this.eventEmitter);
+        this.eventEmitter.emit("valueChanged", event, local, op ?? null, this.eventEmitter);
```
Note that the typings suggest `null`, but we were actually possibly emitting `undefined` as per this local typing. Not sure which is really correct, but generally we lean towards `undefined`.

Commit:
Fix consensus dds

Commit:
Fix task manager

Commit:
Fix sequence copy of mapKernel

Commit:
Fix react package

Commit:
Merge branch 'main' into typed-events

Commit:
Change event type transformers

Commit:
Propoagate new EventThis signature

Commit:
Various fixes

arinwt commented:
Bump! - I merged latest main branch to this + fixed a bunch of build errors in the runtime/dds packages.

I also changed the way the event transforming works by adding `EventLiteral<T>`.

Now you can use either array or function declaration, and use `EventThis` as a `this` placeholder for either. Additionally, you can use normal types in most cases, but if you want to handle complex/generic types, you can specify them as `EventLiteral<T>` to essentially bypass the `EventThis` (and any other future transforms) checks.

Below for example, ripped from the type tests:
```ts
interface IComplexEvents<T> {
    maybe: [
        complex: EventLiteral<Complex<T>>,
        thisOne: EventThis,
        n: number,
    ];
    maybeNot(
        complex: EventLiteral<Complex<T>>,
        thisOne: EventThis,
        n: number,
    ): void;
}
export function checkComplex<T>(emitter: IEventEmitter<IComplexEvents<T>>, complex: Complex<T>) {
    emitter.emit("maybe", complex, emitter, 1);
    emitter.emit("maybeNot", complex, emitter, 2);
    emitter.on("maybe", (complex, thisOne, n) => {
        expectType<Complex<T>>(complex);
        expectType<IEventEmitter<IComplexEvents<T>>>(thisOne);
        expectType<number>(n);
    });
    emitter.on("maybeNot", (complex, thisOne, n) => {
        expectType<Complex<T>>(complex);
        expectType<IEventEmitter<IComplexEvents<T>>>(thisOne);
        expectType<number>(n);
    });
}
```

Commit:
Fix routes to work with latest typings

Commit:
Change events typigns

Commit:
Update typescript version

Commit:
Update package-locks for new typescript version

Commit:
Fix sharedObject using new TS version

arinwt commented:
Also bump TypeScript version from ~4.1.3 to ~4.3.2
